File: 2024/Python/Day4/solutionA.py
def countXMAS(grid):
    word = "XMAS"
    rows = len(grid)
    cols = len(grid[0])
    count = 0

    directions = [
        (0, 1), # right
        (0, -1), # left
        (1, 0), # down
        (-1, 0), # up
        (1, 1), # down-right
        (-1, -1), # up-left
        (1, -1), # down-left
        (-1, 1), # up-right
    ]

    def checkDirection(row, col, dr, dc):
        for i in range(len(word)):
            nr = row + i * dr 
            nc = col + i * dc 
            if nr < 0 or nr >= rows or nc < 0 or nc >= cols or grid[nr][nc] != word[i]:
                return False
        return True

    for r in range(rows):
        for c in range(cols):
            for dr, dc in directions:
                if checkDirection(r, c, dr, dc):
                    count += 1

    return count 


grid = []
with open("../../data/day4.txt") as file:
    for line in file:
        grid.append(line.rstrip('\n'))

result = countXMAS(grid)
print(result)

File: 2024/JS/Day4/solutionA.js
const fs = require("fs");

function countXMAS(grid) {
  const word = "XMAS";
  const rows = grid.length;
  const cols = grid[0].length;
  let count = 0;

  // Directions (right, left, down, up, diagonals)
  const directions = [
    { r: 0, c: 1 }, // right
    { r: 0, c: -1 }, // left
    { r: 1, c: 0 }, // down
    { r: -1, c: 0 }, // up
    { r: 1, c: 1 }, // down-right diagonal
    { r: -1, c: -1 }, // up-left diagonal
    { r: 1, c: -1 }, // down-left diagonal
    { r: -1, c: 1 }, // up-right diagonal
  ];

  // check if the word can be found in the given direction
  function checkDirection(r, c, dr, dc) {
    for (let i = 0; i < word.length; i++) {
      const nr = r + dr * i;
      const nc = c + dc * i;
      if (
        nr < 0 ||
        nr >= rows ||
        nc < 0 ||
        nc >= cols ||
        grid[nr][nc] !== word[i]
      ) {
        return false;
      }
    }
    return true;
  }

  // Loop over each cell in the grid
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      for (const dir of directions) {
        if (checkDirection(r, c, dir.r, dir.c)) {
          count++;
        }
      }
    }
  }

  return count;
}

fs.readFile("../../data/day4.txt", (err, data) => {
  let input = data.toString();
  console.log(countXMAS(input.split("\n")));
});
